SPMN

c0derzer0 opened the pull request:
@alejandromolinaml 

I have implemented structure learning algorithm for Sum-Product-Max Networks. Here is a link to the paper that describes the algorithm. [Sum-Product-Max Networks for Tractable Decision Making](https://www.ijcai.org/Proceedings/16/Papers/264.pdf)

Below is a summary of changes 

Code additions :

- SPMN.py and SPMNHelper.py implement the structure learning algorithm
- SPMNDataUtil.py provides data pre-processing methods for SPMN
- MEU.py calculates Maximum Expected Utility
- Added Max node in Base.py
- Added validity check for SPMNs

Code Modifications:
I feel below are the most important changes I made

- [Adding scope parameter for Context class](https://github.com/SPFlow/SPFlow/commit/ac65ca9e8d1faec98e514647f20296ae7d711652)
- [Adding condition for single valued variables in Histogram](https://github.com/SPFlow/SPFlow/commit/6346487bd30bf73a161bc8449055f104ab0dbae2)
- [Modify assert statement for comparison of length of scope with no. of columns](https://github.com/SPFlow/SPFlow/commit/e154d815572e07d2c85d9f5f0094a09b0ff732d1)

Please review and let me know your feedback.

Additionally, here is link to [SPMN](https://github.com/c0derzer0/SPMN) that provides sample data sets and instructions on getting started with SPMN module.

pranavsubramani commented on README.md:
````diff
@@ -480,6 +480,9 @@ this produces the output:
 ```
+### Using SPMN module
+
+* See [SPMN](https://github.com/c0derzer0/SPMN) for sample datasets and instructions for using SPMN module
````
If possible, please attach a set of instructions to the base of this library as well about a sample test case and how to run it + expected output

c0derzer0 replied:
I have a sample test case, running instructions and output in the link given. I can put that in here.

pranavsubramani commented on src/spn/algorithms/Inference.py:
```diff
@@ -39,6 +39,21 @@ def prod_likelihood(node, children, data=None, dtype=np.float64):
     assert llchildren.dtype == dtype
     return np.prod(llchildren, axis=1).reshape(-1, 1)
 
+#added
```
please use PEP8 standards for comment strings (also the line here is unnecessary)

c0derzer0 replied:
Yeah. I can make those changes

pranavsubramani commented on src/spn/algorithms/Inference.py:
```diff
@@ -39,6 +39,21 @@ def prod_likelihood(node, children, data=None, dtype=np.float64):
+def max_log_likelihood(node, children, data=None, dtype=np.float64):
+    llchildren = np.concatenate(children, axis=1)
+    assert llchildren.dtype == dtype
+    #print("node and llchildren", (node, llchildren))
```
you may want to use the python logging library, so a possible rewriting of this could be to use the logging library as opposed to prints.

c0derzer0 replied:
Sure. That's a good suggestion. I'll use the logging library for next revision

pranavsubramani commented on src/spn/algorithms/Inference.py:
```diff
@@ -39,6 +39,21 @@ def prod_likelihood(node, children, data=None, dtype=np.float64):
+#added
+def max_log_likelihood(node, children, data=None, dtype=np.float64):
+    llchildren = np.concatenate(children, axis=1)
+    assert llchildren.dtype == dtype
```
I'm a little confused as to the necessity of this assert -- isn't it just better to explicitly type caste `llchildren` to the type you'd like it to be? Is there a reason you have chosen to do this?

c0derzer0 replied:
I used it to be consistent with other methods already implemented for sum and prod likelihoods. I kept the template and just changed the return statements.

pranavsubramani replied:
Okay, I see, I think this should be changed everywhere then, since the assert is sort of meaningless.

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+import collections
+
+
+def merge_input_vals(l):
```
What is `l` supposed to be over here, please refrain from using single character variable names.

c0derzer0 replied:
Thanks for pointing it out. It is supposed to be parent_result. I have used the method from MPE.py module as is. You may want to review it there as well.

pranavsubramani replied:
Ah okay, maybe we can change the variable name to be something like `parent_result` or something meaningful?

c0derzer0 replied:
It concatenates `parent_result`, so I think may be using that name makes sense.

pranavsubramani commented on src/spn/algorithms/LearningWrappers.py:
```diff
@@ -113,6 +113,7 @@ def learn_mspn(
     rows="kmeans",
     min_instances_slice=200,
     threshold=0.3,
+    initial_scope=None,
```
What is the need to have an initial scope?

c0derzer0 replied:
For SPMN implementation, we iterate through a partial order of variables and call these functions in each iteration and merge the spns learned. For example, let `V` represent R.V's and` D `represent Decision variables, and  `[[V0,V1], [D0],[V2,V3,V4], [V5,V6]]` be the partial order, so we learn spns on `[V0,V1];scope=[0,1]`, next on [`V2,V3,V4]; scope[2,3,4]` and so on. Hence, the need for initial scope.

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+def merge_input_vals(l):
+    return np.concatenate(l)
+
+def meu_max(node, parent_result, data=None, lls_per_node=None, rand_gen=None):
```
Why does there have to be a `rand_gen` over here, it does not appear to be used in this function?

c0derzer0 replied:
Even here, I have taken the template from mpe_sum and added instructions for returning `decision_values` for that decision node. We may have to review mpe_sum and mpe_prod from MPE.py as well.

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+    for i, c in enumerate(node.children):
+        children_row_ids[c] = parent_result[max_child_branches == i]
+    # print("node and max_child_branches", (node.var_name, max_child_branches))
+    ## decision values at each node
```
change the `##` to `#`

pranavsubramani commented on src/spn/algorithms/MPE.py:
```diff
@@ -71,6 +71,8 @@ def get_mpe_top_down_leaf(node, input_vals, data=None, mode=0):
 _node_bottom_up_mpe = {}
 _node_bottom_up_mpe_log = {Sum: sum_log_likelihood, Product: prod_log_likelihood}
 
+def get_node_funtions():
```
If this is always of fixed size, maybe we return a tuple instead?

c0derzer0 replied:
Right. I'll have to make that change

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+def merge_input_vals(l):
+    return np.concatenate(l)
+
+def meu_max(node, parent_result, data=None, lls_per_node=None, rand_gen=None):
```
Could you add a couple of comments about what the function aims to perform?

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+def meu(node, input_data, node_top_down_meu=_node_top_down_meu, node_bottom_up_meu=_node_bottom_up_meu, in_place=False):
+    valid, err = is_valid(node)
+    assert valid, err
+    if in_place:
```
Sorry, I'm a little confused as to what the `in_place` is supposed to do here? why not just explicitly cast to an `np.array` and have an assert on the top to ensure that it is in fact a list, so we don't get any ugly stack traces of converting to an `np.array` breaking (or `np.ndarray`)

c0derzer0 replied:
Again, I left `in_place` in place to be consistent with `mpe` function in MPE.py. And yeah, having an assert and explicitly casting to `np.array` seems to be a good alternative.

pranavsubramani commented on src/spn/algorithms/MEU.py:
```diff
@@ -0,0 +1,162 @@
+def merge_rows_for_decisions(all_decisions=None):
+    """
+    merges different values of same key into one key:value pair
+    :param all_decisions: takes the dictionary of decisions with key as max_node name and values as numpy array with first column as row number of instance, second column as decision.
```
since you've mentioned the type of `:param all_decisions:`, may I suggest adding another line which says `:type all_decisions:` which details the type as well.

c0derzer0 replied:
Nice, that breaks it down well. Yep, I'll add that.

c0derzer0 commented:
Hey Alejandro,

Yeah, I have made those changes. I am also making more changes to the algorithm and refactored some code. We have also implemented EM for SPMNs. I will send a new pull request once we are done with the changes.
